main/lsp/requests/code_action.cc

```diff
@@ -29,8 +29,229 @@ vector<unique_ptr<TextDocumentEdit>> getEdits(const LSPConfiguration &config, co
+                return;
+            }
+        }
+        // block any method not starting with /[a-zA-Z0-9_]+/. This blocks operator overloads.
+        if (!isalnum(oldName[0]) && oldName[0] != '_') {
+            error = fmt::format("The `{}` method cannot be extracted to a module.", oldName);
+            invalid = true;
+        }
+    }
+
+    ~MethodCallSiteRenamer() {}
+    void rename(unique_ptr<core::lsp::QueryResponse> &response) override {
+        if (invalid) {
+            return;
+        }
+        auto loc = response->getLoc();
+
+        // If we're renaming the exact same place twice, silently ignore it. We reach this condition when we find the
+        // same method send through multiple definitions (e.g. in the case of union types)
+        auto it = edits.find(loc);
+        if (it != edits.end()) {
+            return;
+        }
+
+        auto source = loc.source(gs);
+        if (!source.has_value()) {
+            return;
+        }
+        string newsrc;
+        if (auto sendResp = response->isSend()) {
+            edits[sendResp->receiverLoc] = newName;
+        }
+    }
+    void addSymbol(const core::SymbolRef symbol) override {
+        if (symbol.isMethod()) {
+            addSubclassRelatedMethods(gs, symbol.asMethodRef(), getQueue());
+        } else {
+            getQueue()->tryEnqueue(symbol);
+        }
+    }
+}; // CallSiteRenamer
+
+vector<unique_ptr<TextEdit>> moveMethod(const LSPConfiguration &config, const core::GlobalState &gs,
+                                        const core::lsp::DefinitionResponse *definition,
+                                        LSPTypecheckerInterface &typechecker, string_view newModuleName) {
+    auto moduleStart = fmt::format("module {}\n  extend T::Sig\n  ", newModuleName);
+    auto moduleEnd = "\nend";
+
+    auto fref = definition->termLoc.file();
+    auto &file = fref.data(gs);
+
+    auto trees = typechecker.getResolved({fref});
+    auto &rootTree = trees[0].tree;
+    auto beginLoc = core::Loc::offset2Pos(fref.data(gs), rootTree.loc().beginPos());
+
+    auto pos2Loc = [&](const auto &pos) {
+        auto loc = core::Loc::offset2Pos(fref.data(gs), pos);
+        return make_unique<Position>(loc.line - 1, loc.column - 1);
+    };
+
+    auto range2Loc = [&](const auto &pos) {
+        return make_unique<Range>(pos2Loc(pos.beginPos()), pos2Loc(pos.endPos()));
+    };
+
+    auto topOfTheFile = file.getLine(beginLoc.line);
+    auto sigAndMethodLocs = methodLocs(gs, rootTree, definition->symbol, fref);
+    if (!sigAndMethodLocs.has_value()) {
+        return {};
+    }
+    auto [sigLoc, methodLoc] = sigAndMethodLocs.value();
+    auto methodSource = copyMethodSource(gs, sigLoc, methodLoc, fref);
+
+    auto range = make_unique<Range>(make_unique<Position>(beginLoc.line - 1, 0),
+                                    make_unique<Position>(beginLoc.line - 1, topOfTheFile.length()));
```

Comment:
Why are we replacing anything?

Shouldn't the range used for the replacement always just be an empty range, to effectively insert at the beginning of the file?

I would have expected that we could do

```cpp
auto range = Range::fromLoc(rootTree.loc().copyWithZeroWidth());
```

and delete most of the other code here. In particular I'm of the opinion that we should never construct raw `Range` and `Position` nodes manually, preferring instead to operate exclusively on `Loc`'s and convert to them only just before we need to respond to the LSP query.

Reply:
If we take this approach, it also means that our replacement doesn't have to include `topOfTheFile` in the format string.